### PR TITLE
Give algorithm for number of elements of runtime-sized array

### PR DESCRIPTION
**Pull request by dneto0**
Fixes: #3932

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -8627,6 +8632,36 @@ where compatibility is defined by the following table.
...
+
+<blockquote algorithm="element count of runtime-sized array">
+* Let |T| be the [=store type=] for a [=storage buffer=] variable,
+    where |T| is a [=runtime-sized=] array type or contains a runtime-sized array type.
+* Let |EBS| be the
+    [=effective buffer binding size=]
+    for the {{GPUBufferBinding}} bound to the pipeline binding address corresponding to the storage buffer variable.
+* Then <dfn noexport>NRuntime</sub></dfn>, i.e.
+    the number of elements in the runtime-sized array,
+    is the largest integer such that [=SizeOf=](|T|) &le; |EBS|.
```
Comment (jrprice) [inline wgsl/index.bs:8648]: If the runtime-sized array is inside a structure, you need to take the byte-offset of that struct member into account.

**Review by jrprice — CHANGES_REQUESTED** [review]

File: wgsl/index.bs (hunk 1 above)
Comment (dneto0) [reply wgsl/index.bs:8648]: This does.  |T| is not the runtime-array, it's the store type of the whole variable.
The sizeof equation here takes the whole recursive size computation into account.

Comment (jrprice) [reply wgsl/index.bs:8648]: Ah yes, sorry.

**Comment by dneto0** [comment]
But it does point out that a few examples would help.  I'll log that.

**Comment by dneto0** [comment]
Rebased, and:
- elaborated the algorithm more, to call out the fact you need to take the array offset into account, and
- added examples.

File: wgsl/index.bs [hunk 2]
```diff
@@ -8627,12 +8632,126 @@ where compatibility is defined by the following table.
...
+
+In general, the [=NRuntime=] for a runtime-size array is:
+<blockquote>
+[=truncate=]((EBBS &minus; *array_offset*) &divide; *array_element_size*), where:
+* EBBS is the [=effective buffer binding size=] associated with the variable,
+* *array_offset* is the byte offset of the runtime-sized array within the store type of the variable.
+    * It is zero if the store type is the runtime-sized array itself.
+    * Otherwise the store type is a structure, and its last member is the runtime-sized array.
+        In this case *array_offset* is the [=OffsetOfMember|byte offset of that member=] within the structure.
+* *array_element_size* is the [=SizeOf|size of=] the element type of the runtime-sized array.
```
Comment (jrprice) [inline wgsl/index.bs:8660]: The array element stride may be larger than the `SizeOf` the element type (e.g. for `vec3`).

**Review by jrprice — CHANGES_REQUESTED** [review]

File: wgsl/index.bs (hunk 2 above)
Comment (dneto0) [reply wgsl/index.bs:8660]: of course. Thanks!

File: wgsl/index.bs [hunk 3]
```diff
@@ -8627,12 +8632,126 @@ where compatibility is defined by the following table.
...
+</table>
+</div>
+
 # Language Extensions # {#language-extensions}
 
 The WGSL language is expected to evolve over time.
 
 An <dfn noexport>extension</dfn> is a named grouping of a coherent
-set of modifications to the WGSL specification, consisting of any combination of:
+gset of modifications to the WGSL specification, consisting of any combination of:
```
Comment (alan-baker) [inline wgsl/index.bs:8754]: Added a typo.
```suggestion
set of modifications to the WGSL specification, consisting of any combination of:
```

File: wgsl/index.bs [hunk 4]
```diff
@@ -8627,12 +8632,126 @@ where compatibility is defined by the following table.
 See the [$validating GPUProgrammableStage|WebGPU API$]
 specification for interface validation requirements.
 
+### Buffer Binding Determines Runtime-Sized Array Element Count ### {#buffer-binding-determines-runtime-sized-array-element-count}
```
Comment (alan-baker) [inline wgsl/index.bs:8635]: In WGSL the number of elements is available using the `arrayLength` built-in function. It seems remiss not to link to it in this section. Perhaps the description for arrayLength should also link to this section.

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: wgsl/index.bs (hunk 4 above)
Comment (dneto0) [reply wgsl/index.bs:8635]: Line 12622 updates the arrayLength description to point at NRuntime, whose definition is in the new section.

> It seems remiss not to link to it in this section.

Will add a link to arrayLength from this section.

File: wgsl/index.bs (hunk 2 above)
Comment (dneto0) [reply wgsl/index.bs:8660]: Fixed. PTAL.

File: wgsl/index.bs (hunk 4 above)
Comment (dneto0) [reply wgsl/index.bs:8635]: Fixed PTAL.

**Review by alan-baker — APPROVED** [review]
Nit

**Review by jrprice — APPROVED** [review]

**Comment by dneto0** [comment]
> Nit

I checked with @alan-baker in a side channel.  He meant to delete this.
Will land it.

**Pull request merged**